core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -117,15 +120,18 @@ private[kafka] object LogValidator extends Logging {
                                                    isFromClient: Boolean): ValidationAndOffsetAssignResult = {
     val sizeInBytesAfterConversion = AbstractRecords.estimateSizeInBytes(toMagicValue, offsetCounter.value,
       CompressionType.NONE, records.records)
+    val startOffset = offsetCounter.value
 
     val (producerId, producerEpoch, sequence, isTransactional) = {
       val first = records.batches.asScala.head
       (first.producerId, first.producerEpoch, first.baseSequence, first.isTransactional)
     }
 
     val newBuffer = ByteBuffer.allocate(sizeInBytesAfterConversion)
-    val builder = MemoryRecords.builder(newBuffer, toMagicValue, CompressionType.NONE, timestampType,
+    var builder = MemoryRecords.builder(newBuffer, toMagicValue, CompressionType.NONE, timestampType,
       offsetCounter.value, now, producerId, producerEpoch, sequence, isTransactional, partitionLeaderEpoch)
+    if (isFromClient)
+      builder = builder.withTime(time)
```

Comment:
It seems that the only places where we use `withTime` are in this class. I wonder if it's worth adding the functionality to `MemoryRecordsBuilder` or if it's better to keep that class simple and implement it elsewhere (either this class or some helper).